### PR TITLE
[nnc] Make our exceptions c10::Errors, get C++ stacktraces

### PR DESCRIPTION
**Pull request by bertmaher**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64332

With this diff, if a compiler bug occurs (unlikely, I know!) we'll be able to get a c++ stacktrace leading to the exception, rather than just a terse message.  E.g.,
```
RuntimeError: UNSUPPORTED DTYPE
Exception raised from compilation_error at ../torch/csrc/jit/tensorexpr/exceptions.h:32 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x6b (0x7f966659b2eb in /fsx/users/bertrand/c\
onda/envs/pytorch/lib/python3.8/site-packages/torch/lib/libc10.so)
frame #1: <unknown function> + 0x376f099 (0x7f966a195099 in /fsx/users/bertrand/conda/envs/pytorch/lib/python3.8/site-packages/torch/lib/libtorch_cuda.so)
frame #2: <unknown function> + 0x3763bf5 (0x7f966a189bf5 in /fsx/users/bertrand/conda/envs/pytorch/lib/python3.8/site-packages/torch/lib/libtorch_cuda.so)
frame #3: torch::jit::tensorexpr::CudaCodeGen::Initialize() + 0xdd8 (0x7f966a193368 in /fsx/users/bertrand/conda/envs/pytorch/lib/python3.8/site-packages/torch/lib/libtorch_cuda\
.so)
```

Differential Revision: [D30745610](https://our.internmc.facebook.com/intern/diff/D30745610)

### CODE REVIEW

**Review by ZolotukhinM — APPROVED** [review]
Awesome! You probably need to update this place in vectorizer to fix the failing test:
https://github.com/pytorch/pytorch/blob/6bb4b5d150ab51ed15d15ed270471848bb84d4e3/torch/csrc/jit/tensorexpr/loopnest.cpp#L479

**Comment by navahgar** [comment]
This is awesome. Great work. I was planning to look into this. Thanks for figuring this out.

**Comment by navahgar** [comment]
I'm going to wait for this to land to add `buildErrorMessage` to them.

File: torch/csrc/jit/tensorexpr/exceptions.h [hunk 1]
```diff
@@ -26,61 +26,73 @@ namespace torch {
...
  public:
-  explicit unsupported_dtype() : std::runtime_error("UNSUPPORTED DTYPE") {}
+  explicit compilation_error(const std::string& err)
+      : c10::Error(
+            {
+                __func__,
+                __FILE__,
+                static_cast<uint32_t>(__LINE__),
+            },
+            err) {}
```
Comment (navahgar) [inline torch/csrc/jit/tensorexpr/exceptions.h:38]: Actually, now that you have made this a single base class from which all the classes below derive, if you can replace `err` with `buildErrorMessage(err)`, that will take care of adding the fuser turn off info as well.

**Review by navahgar — APPROVED** [review]

**Comment by bertmaher** [comment]
@bertmaher has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30745610).

**Comment by bertmaher** [comment]
@bertmaher has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30745610).

**Pull request closed without merging**